File: app/src/main/java/com/hsilighting/pagify/ui/components/PagePicker.kt
package com.hsilighting.pagify.ui.components

import android.graphics.Bitmap
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.unit.dp
import androidx.compose.foundation.Image
import com.hsilighting.pagify.core.PageSize

/**
 * Choosing pages out of a document.
 *
 * Takes its pages as lambdas rather than a document, because it is used against
 * two different ones: the document being read, when choosing what to export, and
 * a file just opened, when choosing what to import from it. Neither knows about
 * the other.
 *
 * Selection is a *list*, not a set. The order pages are tapped in is the order
 * they come out in — "give me page 3, then page 1" is a thing somebody can ask
 * for, and quietly sorting it hands them a different document.
 */
@Composable
fun PagePicker(
    pageCount: Int,
    /** What the button says, e.g. "Export 3 pages". */
    confirmLabel: (Int) -> String,
    onConfirm: (List<Int>) -> Unit,
    onCancel: () -> Unit,
    pageSizeProvider: suspend (Int) -> PageSize?,
    renderer: suspend (pageIndex: Int, zoom: Float) -> Bitmap?,
    modifier: Modifier = Modifier,
) {
    var chosen by remember(pageCount) { mutableStateOf(listOf<Int>()) }

    Column(modifier.padding(horizontal = 16.dp)) {
        Row(
            Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(
                text = if (chosen.isEmpty()) {
                    "Tap the pages you want"
                } else {
                    "${chosen.size} of $pageCount chosen"
                },
                style = MaterialTheme.typography.titleMedium,
            )
            TextButton(
                onClick = {
                    chosen = if (chosen.size == pageCount) {
                        emptyList()
                    } else {
                        (0 until pageCount).toList()
                    }
                },
            ) { Text(if (chosen.size == pageCount) "None" else "All") }
        }

        LazyVerticalGrid(
            columns = GridCells.Adaptive(minSize = 108.dp),
            horizontalArrangement = Arrangement.spacedBy(10.dp),
            verticalArrangement = Arrangement.spacedBy(10.dp),
            // `fill = false` so a two-page document does not stretch the sheet to
            // full height with empty space under the grid.
            modifier = Modifier
                .weight(1f, fill = false)
                .padding(vertical = 12.dp),
        ) {
            items(items = (0 until pageCount).toList(), key = { it }) { index ->
                val order = chosen.indexOf(index)
                PickerCell(
                    index = index,
                    // The position in the chosen list, not a tick: with order
                    // mattering, "3rd" is the fact somebody needs to see, and a
                    // tick would hide it.
                    order = if (order >= 0) order + 1 else null,
                    onTap = {
                        chosen = if (order >= 0) chosen - index else chosen + index
                    },
                    pageSizeProvider = pageSizeProvider,
                    renderer = renderer,
                )
            }
        }

        Row(
            Modifier
                .fillMaxWidth()
                .padding(bottom = 12.dp),
            horizontalArrangement = Arrangement.End,
        ) {
            TextButton(onClick = onCancel) { Text("Cancel") }
            TextButton(
                enabled = chosen.isNotEmpty(),
                onClick = { onConfirm(chosen) },
            ) { Text(confirmLabel(chosen.size)) }
        }
    }
}

@Composable
private fun PickerCell(
    index: Int,
    order: Int?,
    onTap: () -> Unit,
    pageSizeProvider: suspend (Int) -> PageSize?,
    renderer: suspend (pageIndex: Int, zoom: Float) -> Bitmap?,
) {
    var bitmap by remember(index) { mutableStateOf<Bitmap?>(null) }
    var ratio by remember(index) { mutableStateOf(DEFAULT_PAGE_RATIO) }

    LaunchedEffect(index) {
        val size = pageSizeProvider(index) ?: return@LaunchedEffect
        if (size.widthPoints > 0f && size.heightPoints > 0f) {
            ratio = size.widthPoints / size.heightPoints
        }
        bitmap = renderer(index, THUMBNAIL_ZOOM)
    }

    Column(horizontalAlignment = Alignment.CenterHorizontally) {
        Box(
            Modifier
                .fillMaxWidth()
                .aspectRatio(ratio)
                .clip(RoundedCornerShape(6.dp))
                .background(Color.White)
                .border(
                    width = if (order != null) 3.dp else 1.dp,
                    color = if (order != null) {
                        MaterialTheme.colorScheme.primary
                    } else {
                        MaterialTheme.colorScheme.outlineVariant
                    },
                    shape = RoundedCornerShape(6.dp),
                )
                .clickable(onClick = onTap),
        ) {
            bitmap?.let {
                Image(
                    bitmap = it.asImageBitmap(),
                    contentDescription = null,
                    contentScale = ContentScale.Fit,
                    modifier = Modifier.fillMaxWidth(),
                )
            }
            if (order != null) {
                Box(
                    Modifier
                        .align(Alignment.TopEnd)
                        .padding(4.dp)
                        .size(22.dp)
                        .clip(CircleShape)
                        .background(MaterialTheme.colorScheme.primary),
                    contentAlignment = Alignment.Center,
                ) {
                    if (order > MOST_SHOWN_AS_A_NUMBER) {
                        Icon(
                            Icons.Filled.Check,
                            contentDescription = null,
                            tint = MaterialTheme.colorScheme.onPrimary,
                            modifier = Modifier.size(14.dp),
                        )
                    } else {
                        Text(
                            text = "$order",
                            style = MaterialTheme.typography.labelSmall,
                            color = MaterialTheme.colorScheme.onPrimary,
                        )
                    }
                }
            }
        }
        Text(
            text = "${index + 1}",
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.padding(top = 4.dp),
        )
    }
}

/** A4 upright, for the moment before the real page size is known. */
private const val DEFAULT_PAGE_RATIO = 595f / 842f

/** Small enough to be quick, large enough to tell two pages apart. */
private const val THUMBNAIL_ZOOM = 0.35f

/**
 * Past this, the badge shows a tick instead of a number.
 *
 * Two digits fit; three do not, and a "12" squeezed to illegibility is worse than
 * a tick. Somebody selecting a hundred pages is selecting all of them anyway.
 */
private const val MOST_SHOWN_AS_A_NUMBER = 99

File: ios/Sources/PagePicker.swift
import SwiftUI

/// Choosing pages out of a document.
///
/// Takes a document rather than the reader's model, because it is used against
/// two of them: the one being read, when choosing what to export, and a file
/// just opened, when choosing what to import from it. Neither knows about the
/// other, and only one of them is the document the reader is editing.
///
/// Selection is a *list*, not a set. The order pages are tapped in is the order
/// they come out in — "give me page 3, then page 1" is a thing somebody can ask
/// for, and quietly sorting it hands them a different document. `exportPages`
/// and `importPages` both take the list as given, so the position each chosen
/// page shows is what will actually happen.
struct PagePicker: View {
    let document: PagifyDocument

    /// Changes when the pages stop being what is drawn, so the thumbnails are
    /// re-rendered. Left at zero for a file opened only to import from: nothing
    /// here can edit it.
    var revision: Int = 0

    /// What the button says, e.g. "Export 3 pages".
    let confirmLabel: (Int) -> String
    let onConfirm: ([Int]) -> Void
    let onCancel: () -> Void

    @Environment(\.colorScheme) private var scheme
    @State private var chosen: [Int] = []

    private var pageCount: Int { document.pageCount }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            header
            grid
            footer
        }
        .padding(.horizontal, 16)
        .background(PagifyColor.background(scheme))
    }

    private var header: some View {
        HStack {
            Text(chosen.isEmpty ? "Tap the pages you want"
                                : "\(chosen.count) of \(pageCount) chosen")
                .font(.headline)

            Spacer()

            Button(chosen.count == pageCount ? "None" : "All") {
                chosen = chosen.count == pageCount ? [] : Array(0..<pageCount)
            }
        }
        .padding(.top, 16)
    }

    private var grid: some View {
        ScrollView {
            LazyVGrid(columns: [GridItem(.adaptive(minimum: 108), spacing: 10)], spacing: 10) {
                ForEach(Array(0..<pageCount), id: \.self) { index in
                    let position = chosen.firstIndex(of: index).map { $0 + 1 }
                    // The rail's cell, which already draws a page at its own
                    // shape and rings the one that is picked out. What "picked
                    // out" means is the only thing that differs here, and it is
                    // the badge that says so.
                    PageThumbnail(document: document,
                                  index: index,
                                  revision: revision,
                                  isCurrent: position != nil,
                                  onSelect: { toggle(index) })
                        .overlay(alignment: .topTrailing) {
                            if let position { badge(position) }
                        }
                        // Replacing the rail's own label, which says "Go to
                        // page" — nothing here goes anywhere, and the position
                        // is the part a badge cannot read out.
                        .accessibilityLabel(position.map {
                            "Page \(index + 1), chosen at position \($0)"
                        } ?? "Page \(index + 1)")
                }
            }
            .padding(.vertical, 12)
        }
    }

    private var footer: some View {
        HStack {
            Spacer()
            Button("Cancel", action: onCancel)
            Button(confirmLabel(chosen.count)) { onConfirm(chosen) }
                .buttonStyle(.borderedProminent)
                .disabled(chosen.isEmpty)
        }
        .padding(.bottom, 16)
    }

    /// The position in the chosen list, not a tick.
    ///
    /// With order mattering, "3rd" is the fact somebody needs to see, and a tick
    /// would hide it. It never takes the tap: the whole cell is the target, and
    /// a badge that swallowed touches would make a chosen page impossible to
    /// unchoose.
    private func badge(_ position: Int) -> some View {
        ZStack {
            Circle().fill(PagifyColor.primary(scheme))
            if position > Self.mostShownAsANumber {
                Image(systemName: "checkmark")
                    .font(.system(size: 11, weight: .bold))
            } else {
                Text("\(position)").font(.caption2.weight(.semibold))
            }
        }
        .foregroundStyle(PagifyColor.onPrimary(scheme))
        .frame(width: 22, height: 22)
        .padding(4)
        .allowsHitTesting(false)
    }

    private func toggle(_ index: Int) {
        if let at = chosen.firstIndex(of: index) {
            chosen.remove(at: at)
        } else {
            chosen.append(index)
        }
    }

    /// Past this, the badge shows a tick instead of a number.
    ///
    /// Two digits fit; three do not, and a "123" squeezed to illegibility is
    /// worse than a tick. Somebody selecting a hundred pages is selecting all of
    /// them anyway.
    private static let mostShownAsANumber = 99
}
